### PR TITLE
Optimize Quantity.parse to avoid regex overhead

### PR DESCRIPTION
**Pull request by schlosna**
## Description
Optimize `Quantity.parse` to avoid regex overhead. While reviewing some JFR profiles from a service using fabric8.io k8s client, I noticed some significant allocations and CPU time spent parsing quantities using regex from interacting with k8s APIs. We can avoid some of these overheads by simplifying the parsing to identify when and where to split quantity and units.

image
image
image

## Type of change

 - [x] Bug fix (non-breaking change which fixes an issue)
 - [ ] Feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to change
 - [ ] Chore (non-breaking change which doesn't affect codebase;
   test, version modification, documentation, etc.)

## Checklist
 - [x] Code contributed by me aligns with current project license: [Apache 2.0](https://www.apache.org/licenses/LICENSE-2.0)
 - [x] I Added [CHANGELOG](https://github.com/fabric8io/kubernetes-client/blob/master/CHANGELOG.md) entry regarding this change
 - [x] I have implemented unit tests to cover my changes
 - [ ] I have added/updated the [javadocs](https://www.javadoc.io/doc/io.fabric8/kubernetes-client/latest/index.html) and other [documentation](https://github.com/fabric8io/kubernetes-client/blob/master/doc/CHEATSHEET.md) accordingly
 - [ ] No new bugs, code smells, etc. in [SonarCloud](https://sonarcloud.io/dashboard?id=fabric8io_kubernetes-client) report
 - [ ] I tested my code in Kubernetes
 - [ ] I tested my code in OpenShift

### CODE REVIEW

File: kubernetes-model-generator/kubernetes-model-core/src/main/java/io/fabric8/kubernetes/api/model/Quantity.java [hunk 1]
```diff
@@ -249,17 +260,47 @@ public static Quantity parse(String quantityAsString) {
...
+        case 'k':
+        case 'K':
+        case 'M':
+        case 'G':
+        case 'T':
+        case 'P':
+          return i;
+        default:
+          //noinspection UnnecessaryContinue - satisfy Sonar
+          continue;
```
Comment (schlosna) [inline kubernetes-model-generator/kubernetes-model-core/src/main/java/io/fabric8/kubernetes/api/model/Quantity.java:298]: Without this `default` case, [Sonar flags this as a critical `Add a default case to this switch.`](https://sonarcloud.io/project/issues?resolved=false&types=CODE_SMELL&pullRequest=4992&id=fabric8io_kubernetes-client&open=AYcPRu7YtWsAr5uOwQsf) , though with the `default` case of `continue` IntelliJ then flags it as unnecessary. Defer to your preferences if you'd like an empty `default` case or if the comment suppressing IntelliJ warning is ok with you all.

**Review by rohanKanojia — APPROVED** [review]

**Review by andreaTP — APPROVED** [review]

**Comment by schlosna** [comment]
It looks like the `Sonar Scanner / Sonar (pull_request)` [check is failing](https://github.com/fabric8io/kubernetes-client/actions/runs/4503146389/jobs/7937925204?pr=4992) on `JettyHttpClientTest>OkHttpClientTest.supportedResponseBodyTypes:197`. I don't think this is related to this change, is this a known test flake?

Let me know if there's anything else you need to merge. Thanks!

**Comment by shawkins** [comment]
> I don't think this is related to this change, is this a known test flake?

It's definitely not related, but it's not a known test flake.  Captured as #5002 - there does appear to be a timing issue with the jetty logic.

File: kubernetes-model-generator/kubernetes-model-core/src/main/java/io/fabric8/kubernetes/api/model/Quantity.java (hunk 1 above)
Comment (manusa) [reply kubernetes-model-generator/kubernetes-model-core/src/main/java/io/fabric8/kubernetes/api/model/Quantity.java:298]: Sonar false positives can be disabled by using a `@SuppressWarnings` annotation followed by the rule id e.g. `@SuppressWarnings("java:S107")`

**Review by manusa — APPROVED** [review]
LGTM, thx!

**Review by manusa — APPROVED** [review]

**Pull request merged**